R/r-session.R

```diff
@@ -0,0 +1,306 @@
+
+  ## Run an expr that loads it, in the child process, with error handlers
+  expr <- make_vanilla_script_expr(private$options$func_file,
+                                   private$options$result_file,
+                                   private$options$error)
+  cmd <- paste0(deparse(expr), "\n")
+
+  ## Write this to stdin
+  private$write_for_sure(cmd)
+
+  ## Report back when done
+  report_str <- paste0("DONE", basename(private$options$result_file))
+  private$report_back(200, report_str)
+
+  private$state <- "busy"
+}
+
+rs_wait_for_call <- function(self, private, timeout) {
+  if (private$state %in% c("finished", "ready", "idle")) return()
+
+  pr <- poll(list(private$pipe[[1]]), timeout)[[1]]
+  if (pr == "ready") {
+    if (private$state == "starting") {
+      private$state <- "idle"
+    } else {
+      private$state <- "ready"
+    }
+    invisible(conn_read_lines(private$pipe[[1]], 1))
+  } else {
+    invisible()
+  }
+}
+
+rs_get_result <- function(self, private) {
+  if (private$state != "ready") private$update_state()
+
+  get_my_result <- function() {
+    ## This is artificial...
+    out <- list(
+      status = 0,
+      stdout = "",
+      stderr = "",
+      timeout = FALSE
+    )
+    res <- get_result(out, private$options)
+    private$state <- "idle"
+    unlink(private$options$tmp_files, recursive = TRUE)
+    private$options$tmp_files <- NULL
+    res
+  }
+
+  switch(
+    private$state,
+    "finished" = stop("R session already finished"),
+    "idle" = stop("R session is idle"),
+    "busy" = stop("R session still busy"),
+    "starting" = stop("R session still starting"),
+    "ready" = get_my_result()
+  )
+}
+
+rs_get_running_time <- function(self, private) {
+  now <- Sys.time()
+  c(total = now - private$started_at,
+    current = now - private$fun_started_at)
+}
+
+rs_get_state <- function(self, private) {
+  private$update_state()
+  private$state
+}
+
+rs_finish <- function(self, private, grace) {
+  close(self$get_input_connection())
+  self$poll_io(grace)
+  self$kill()
+  private$state <- "finished"
+  private$fun_started_at <- as.POSIXct(NA)
+}
+
+#' @importFrom processx conn_read_lines
+
+rs__update_state <- function(self, private) {
+  self$wait_for_call(timeout = 0)
+}
+
+rs__report_back <- function(self, private, code, text) {
+  cmd <- paste0(deparse(rs__status_expr(code, text, fd = 3)), "\n")
+  private$write_for_sure(cmd)
+}
+
+rs__write_for_sure <- function(self, private, text) {
+  while (1) {
+    text <- self$write_input(text)
+    if (!length(text)) break;
+    Sys.sleep(.1)
+  }
+}
+
+#' @importFrom processx conn_create_fd conn_write
+
+rs__status_expr <- function(code, text = "", fd = 3) {
+  substitute(
+    {
+      code_ <- code; fd_ <- fd; text_ <- text
+      con <- processx::conn_create_fd(fd_, close = FALSE)
+      data <- paste0(code_, " ", text_, "\n")
+      while (1) {
+        data <- processx::conn_write(con, data)
+        if (!length(data)) break;
+        Sys.sleep(.1)
```

Comment:
I wonder if the sleep time here and on L274 should be a parameter of the object?

Reply:
Could be, but this sleep is actually not vey important. It is a safety net that is almost never needed. Basically it is only needed if the write buffer is full, but the write buffer is at least 4k or 8k, and we don't send messages longer than that, and only ever keep at most one message in the buffer. So it is not really important to make this user configurable, I think.